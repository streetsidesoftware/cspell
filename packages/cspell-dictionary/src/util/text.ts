const regExFirstUpper = /^\p{Lu}\p{M}?\p{Ll}+$/u;
const regExAllUpper = /^(?:\p{Lu}\p{M}?)+$/u;
const regExAllLower = /^(?:\p{Ll}\p{M}?)+$/u;
const regExAccents = /\p{M}/gu;

export function isUpperCase(word: string): boolean {
    return !!word.match(regExAllUpper);
}

export function isLowerCase(word: string): boolean {
    return !!word.match(regExAllLower);
}

export function isFirstCharacterUpper(word: string): boolean {
    return isUpperCase(word.slice(0, 1));
}

export function isFirstCharacterLower(word: string): boolean {
    return isLowerCase(word.slice(0, 1));
}

export function ucFirst(word: string): string {
    return word.slice(0, 1).toUpperCase() + word.slice(1);
}

export function lcFirst(word: string): string {
    return word.slice(0, 1).toLowerCase() + word.slice(1);
}

export function matchCase(example: string, word: string): string {
    if (example.match(regExFirstUpper)) {
        return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase();
    }
    if (example.match(regExAllLower)) {
        return word.toLowerCase();
    }
    if (example.match(regExAllUpper)) {
        return word.toUpperCase();
    }

    if (isFirstCharacterUpper(example)) {
        return ucFirst(word);
    }

    if (isFirstCharacterLower(example)) {
        return lcFirst(word);
    }

    return word;
}

export function removeAccents(text: string): string {
    return text.normalize('NFD').replace(regExAccents, '');
}

export function removeUnboundAccents(text: string): string {
    return text.replace(regExAccents, '');
}
