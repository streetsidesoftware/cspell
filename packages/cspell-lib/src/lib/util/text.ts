import { opConcatMap, opMap, pipe } from '@cspell/cspell-pipe/sync';
import type { TextDocumentOffset, TextOffset } from '@cspell/cspell-types';

import type { Uri } from './IUri.js';
import { binarySearch } from './search.js';
import {
    regExAccents,
    regExAllLower,
    regExAllUpper,
    regExFirstUpper,
    regExIgnoreCharacters,
    regExpCamelCaseWordBreaksWithEnglishSuffix,
    regExWords,
    regExWordsAndDigits,
} from './textRegex.js';
import { toUri } from './Uri.js';
import { scanMap } from './util.js';

export { stringToRegExp } from './textRegex.js';

// CSpell:ignore ings ning gimuy tsmerge

export function splitCamelCaseWordWithOffset(wo: TextOffset): TextOffset[] {
    return splitCamelCaseWord(wo.text).map(
        scanMap<string, TextOffset>((last, text) => ({ text, offset: last.offset + last.text.length }), {
            text: '',
            offset: wo.offset,
        }),
    );
}

/**
 * Split camelCase words into an array of strings.
 */
export function splitCamelCaseWord(word: string): string[] {
    return splitWord(word, regExpCamelCaseWordBreaksWithEnglishSuffix);
}

export function splitWordWithOffset(wo: TextOffset, regExpWordBreaks: RegExp): TextOffset[] {
    return splitWord(wo.text, regExpWordBreaks).map(
        scanMap<string, TextOffset>((last, text) => ({ text, offset: last.offset + last.text.length }), {
            text: '',
            offset: wo.offset,
        }),
    );
}

/**
 * Split camelCase words into an array of strings.
 */
export function splitWord(word: string, regExpWordBreaks: RegExp): string[] {
    return word.split(regExpWordBreaks);
}

/**
 * This function lets you iterate over regular expression matches.
 */
export function match(reg: RegExp, text: string): Iterable<RegExpExecArray> {
    if (!text) return [];
    reg = reg.global ? reg : new RegExp(reg.source, reg.flags + 'g');
    return text.matchAll(reg);
}

export function matchStringToTextOffset(reg: RegExp, text: string): Iterable<TextOffset> {
    return matchToTextOffset(reg, { text, offset: 0 });
}

export function matchToTextOffset(reg: RegExp, t: TextOffset): Iterable<TextOffset> {
    const text = t.text;
    const offset = t.offset;
    // return opMap((m: RegExpExecArray) => ({ text: m[0], offset: offset + m.index }))(match(reg, text));
    return pipe(
        match(reg, text),
        opMap((m) => ({ text: m[0], offset: offset + m.index })),
    );
}

export function* extractLinesOfText(text: string): Iterable<TextOffset> {
    let i = 0;
    for (let j = text.indexOf('\n', i); j >= 0; j = text.indexOf('\n', i)) {
        const end = j + 1;
        yield { text: text.slice(i, end), offset: i };
        i = end;
    }
    yield { text: text.slice(i, text.length), offset: i };
}

/**
 * Extract out whole words from a string of text.
 */
export function extractWordsFromText(text: string): Iterable<TextOffset> {
    return extractWordsFromTextOffset(textOffset(text));
}

/**
 * Extract out whole words from a string of text.
 */
export function extractWordsFromTextOffset(text: TextOffset): Iterable<TextOffset> {
    const reg = new RegExp(regExWords);
    return matchToTextOffset(reg, cleanTextOffset(text));
}

/**
 * Remove Hiragana, Han, Katakana, Hangul characters from the text.
 * @param text
 * @returns the text with the characters removed.
 */
export function cleanText(text: string): string {
    regExIgnoreCharacters.lastIndex = 0;
    if (!regExIgnoreCharacters.test(text)) return text;
    text = text.replace(regExIgnoreCharacters, (match: string) => ' '.repeat(match.length));
    return text;
}

export function cleanTextOffset(text: TextOffset): TextOffset {
    // Do not make a new object if the text is already clean.
    regExIgnoreCharacters.lastIndex = 0;
    if (!regExIgnoreCharacters.test(text.text)) return text;
    return {
        text: cleanText(text.text),
        offset: text.offset,
    };
}

/**
 * Extract out whole words and words containing numbers from a string of text.
 */
export function extractPossibleWordsFromTextOffset(text: TextOffset): Iterable<TextOffset> {
    const reg = new RegExp(regExWordsAndDigits);
    return matchToTextOffset(reg, text);
}

export function extractWordsFromCode(text: string): Iterable<TextOffset> {
    return extractWordsFromCodeTextOffset(textOffset(text));
}

export function extractWordsFromCodeTextOffset(textOffset: TextOffset): Iterable<TextOffset> {
    return pipe(extractWordsFromTextOffset(textOffset), opConcatMap(splitCamelCaseWordWithOffset));
}

export function isUpperCase(word: string): boolean {
    return regExAllUpper.test(word);
}

export function isLowerCase(word: string): boolean {
    return regExAllLower.test(word);
}

export function isFirstCharacterUpper(word: string): boolean {
    return isUpperCase(word.slice(0, 1));
}

export function isFirstCharacterLower(word: string): boolean {
    return isLowerCase(word.slice(0, 1));
}

export function ucFirst(word: string): string {
    return word.slice(0, 1).toUpperCase() + word.slice(1);
}

export function lcFirst(word: string): string {
    return word.slice(0, 1).toLowerCase() + word.slice(1);
}

export function snakeToCamel(word: string): string {
    return word.split('_').map(ucFirst).join('');
}

export function camelToSnake(word: string): string {
    return splitCamelCaseWord(word).join('_').toLowerCase();
}

export function matchCase(example: string, word: string): string {
    if (regExFirstUpper.test(example)) {
        return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase();
    }
    if (regExAllLower.test(example)) {
        return word.toLowerCase();
    }
    if (regExAllUpper.test(example)) {
        return word.toUpperCase();
    }

    if (isFirstCharacterUpper(example)) {
        return ucFirst(word);
    }

    if (isFirstCharacterLower(example)) {
        return lcFirst(word);
    }

    return word;
}

export function textOffset(text: string, offset = 0): TextOffset {
    return { text, offset };
}

export function extractText(textOffset: TextOffset, startPos: number, endPos: number): string {
    const { text, offset: orig } = textOffset;
    const a = Math.max(startPos - orig, 0);
    const b = Math.max(endPos - orig, 0);
    return text.slice(a, b);
}

export function calculateTextDocumentOffsets<T extends TextOffset>(
    uri: string | Uri | URL,
    doc: string,
    wordOffsets: T[],
): (TextDocumentOffset & T)[] {
    const lines = [
        -1,
        ...pipe(
            match(/\n/g, doc),
            opMap((a) => a.index),
        ),
        doc.length,
    ];

    let lastRow = -1;
    let lastOffset = doc.length + 1;
    let lastLineRow = -1;
    let lastLine: TextOffset | undefined;

    function findRowCol(offset: number): [number, number] {
        const row = binarySearch(lines, offset, offset >= lastOffset ? lastRow : undefined);
        const col = offset - lines[Math.max(0, row - 1)];
        lastOffset = offset;
        lastRow = row;
        return [row, col];
    }

    function extractLine(row: number): TextOffset {
        const offset = lines[row - 1] + 1;
        const text = doc.slice(offset, lines[row] + 1);
        return { text, offset };
    }

    function calcLine(row: number): TextOffset {
        const last = lastLineRow === row ? lastLine : undefined;
        lastLineRow = row;
        const r = last ?? extractLine(row);
        lastLine = r;
        return r;
    }

    const _uri = toUri(uri).toString();

    return wordOffsets.map((wo) => {
        const [row, col] = findRowCol(wo.offset);
        return { ...wo, row, col, doc, uri: _uri, line: calcLine(row) };
    });
}

export function removeAccents(text: string): string {
    return text.normalize('NFD').replace(regExAccents, '');
}

export const __testing__ = {
    regExWords,
    regExWordsAndDigits,
};
