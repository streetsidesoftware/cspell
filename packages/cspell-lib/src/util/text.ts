import { xregexp as XRegExp } from 'cspell-util-bundle';
import {Sequence, sequenceFromRegExpMatch } from 'gensequence';
import {binarySearch} from './search';
import { scanMap } from './util';

// CSpell:ignore ings ning gimuy tsmerge

export interface TextOffset {
    text: string;
    offset: number;
}

export interface TextDocumentOffset extends TextOffset {
    uri?: string;
    doc: string;
    row: number;
    col: number;
}

const regExLines = /.*(\r?\n|$)/g;
// const regExIdentifiers = XRegExp('(?:\\p{L}|[0-9_\'])+', 'gi');
const regExUpperSOrIng = XRegExp('(\\p{Lu}+\\\\?[\'’]?(?:s|ing|ies|es|ings|ed|ning))(?!\\p{Ll})', 'g');
const regExSplitWords = XRegExp('(\\p{Ll})(\\p{Lu})', 'g');
const regExSplitWords2 = XRegExp('(\\p{Lu})(\\p{Lu}\\p{Ll})', 'g');
const regExWords = XRegExp("\\p{L}(?:\\\\?['’]\\p{L}|\\p{L})+|\\p{L}", 'g');
const regExIgnoreCharacters = XRegExp('\\p{Hiragana}|\\p{Han}|\\p{Katakana}|[\\u30A0-\\u30FF]|[\\p{Hangul}]', 'g');
const regExFirstUpper = XRegExp('^\\p{Lu}\\p{Ll}+$');
const regExAllUpper = XRegExp('^\\p{Lu}+$');
const regExAllLower = XRegExp('^\\p{Ll}+$');

const regExMatchRegExParts = /^\/(.*)\/([gimuy]*)$/;

const regExAccents = XRegExp('\\p{M}', 'g');

export function splitCamelCaseWordWithOffset(wo: TextOffset): Array<TextOffset> {
    return splitCamelCaseWord(wo.text)
        .map(scanMap<string, TextOffset>(
            (last, text) => ({ text, offset: last.offset + last.text.length }),
            { text: '', offset: wo.offset }
        ));
}

/**
 * Split camelCase words into an array of strings.
 */
export function splitCamelCaseWord(word: string): string[] {
    const wPrime = word.replace(regExUpperSOrIng, s => s[0] + s.substr(1).toLowerCase());
    const separator = '_<^*_*^>_';
    const pass1 = XRegExp.replace(wPrime, regExSplitWords, '$1' + separator + '$2');
    const pass2 = XRegExp.replace(pass1, regExSplitWords2, '$1' + separator + '$2');
    return XRegExp.split(pass2, separator);
}

/**
 * This function lets you iterate over regular expression matches.
 */
export function match(reg: RegExp, text: string): Sequence<RegExpExecArray> {
    return sequenceFromRegExpMatch(reg, text);
}

export function matchStringToTextOffset(reg: RegExp, text: string) {
    return matchToTextOffset(reg, { text, offset: 0 });
}

export function matchToTextOffset(reg: RegExp, text: TextOffset): Sequence<TextOffset> {
    const textOffset = text;
    const fnOffsetMap = offsetMap(textOffset.offset);
    return match(reg, textOffset.text)
        .map(m => fnOffsetMap({ text: m[0], offset: m.index }));
}

export function extractLinesOfText(text: string): Sequence<TextOffset> {
    return matchStringToTextOffset(regExLines, text);
}

/**
 * Extract out whole words from a string of text.
 */
export function extractWordsFromText(text: string): Sequence<TextOffset> {
    return extractWordsFromTextOffset(textOffset(text));
}

/**
 * Extract out whole words from a string of text.
 */
export function extractWordsFromTextOffset(text: TextOffset): Sequence<TextOffset> {
    const reg = XRegExp(regExWords);
    const reg2 = XRegExp(regExWords);
    return matchToTextOffset(reg, text)
        // remove characters that match against \p{L} but are not letters (Chinese characters are an example).
        .map(({ text, offset }) => ({
            text: XRegExp.replace(text, regExIgnoreCharacters, match => ' '.repeat(match.length)),
            offset,
        }))
        .concatMap(wo => matchToTextOffset(reg2, wo))
        .filter(wo => !!wo.text);
}

export function extractWordsFromCode(text: string): Sequence<TextOffset> {
    return extractWordsFromCodeTextOffset(textOffset(text));
}

export function extractWordsFromCodeTextOffset(textOffset: TextOffset): Sequence<TextOffset> {
    return extractWordsFromTextOffset(textOffset)
        .concatMap(splitCamelCaseWordWithOffset);
}

export function isUpperCase(word: string) {
    return !!word.match(regExAllUpper);
}

export function isLowerCase(word: string) {
    return !!word.match(regExAllLower);
}

export function isFirstCharacterUpper(word: string) {
    return isUpperCase(word.slice(0, 1));
}

export function isFirstCharacterLower(word: string) {
    return isLowerCase(word.slice(0, 1));
}

export function ucFirst(word: string) {
    return word.slice(0, 1).toUpperCase() + word.slice(1);
}

export function lcFirst(word: string) {
    return word.slice(0, 1).toLowerCase() + word.slice(1);
}

export function snakeToCamel(word: string) {
    return word.split('_').map(ucFirst).join('');
}

export function camelToSnake(word: string) {
    return splitCamelCaseWord(word).join('_').toLowerCase();
}

export function matchCase(example: string, word: string): string {
    if (example.match(regExFirstUpper)) {
        return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase();
    }
    if (example.match(regExAllLower)) {
        return word.toLowerCase();
    }
    if (example.match(regExAllUpper)) {
        return word.toUpperCase();
    }

    if (isFirstCharacterUpper(example)) {
        return ucFirst(word);
    }

    if (isFirstCharacterLower(example)) {
        return lcFirst(word);
    }

    return word;
}

export function textOffset(text: string, offset: number = 0): TextOffset {
    return { text, offset };
}

export function extractText(textOffset: TextOffset, startPos: number, endPos: number) {
    const { text, offset: orig } = textOffset;
    const a = Math.max(startPos - orig, 0);
    const b = Math.max(endPos - orig, 0);
    return text.slice(a, b);
}

interface OffsetMap {
    offset: number;
}
function offsetMap(offset: number) {
    return <T extends OffsetMap>(xo: T) => ({...(xo as Object), offset: xo.offset + offset }) as T;
}

export function stringToRegExp(pattern: string | RegExp, defaultFlags = 'gim', forceFlags = 'g') {
    if (pattern instanceof RegExp) {
        return pattern;
    }
    try {
        const [, pat, flag] = [...(pattern.match(regExMatchRegExParts) || ['', pattern, defaultFlags]), forceFlags];
        // Make sure the flags are unique.
        const flags = [...(new Set(forceFlags + flag))].join('').replace(/[^gimuy]/g, '');
        if (pat) {
            const regex = new RegExp(pat, flags);
            return regex;
        }
    } catch (e) {
    }
    return undefined;
}

export function calculateTextDocumentOffsets(uri: string, doc: string, wordOffsets: TextOffset[]): TextDocumentOffset[] {
    const lines = [-1, ...match(/\n/g, doc).map(a => a.index), doc.length];

    function findRowCol(offset: number): [number, number] {
        const row = binarySearch(lines, offset);
        const col = offset - lines[Math.max(0, row - 1)];
        return [row, col];
    }

    return wordOffsets
        .map(wo => {
            const [row, col] = findRowCol(wo.offset);
            return { ...wo, row, col, doc, uri };
        });
}

export function removeAccents(text: string) {
    return text.normalize('NFKD').replace(regExAccents, '');
}
