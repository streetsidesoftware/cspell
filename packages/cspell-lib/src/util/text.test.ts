import { splitCamelCaseWord } from './text';
import * as Text from './text';

// cSpell:ignore Ápple DBAs ctrip γάμμα

describe('Util Text', () => {
    test('tests build regexp from string', () => {
        const regEx1 = Text.stringToRegExp('');
        expect(regEx1).toBeUndefined();
        const regEx2 = Text.stringToRegExp('a');
        expect(regEx2!.toString()).toBe('/a/gim');
        const regEx3 = Text.stringToRegExp('/');
        expect(regEx3).not.toBeUndefined();
        expect(regEx3!.toString()).toBe('/\\//gim');
        const regEx4 = Text.stringToRegExp(/abc/);
        expect(regEx4!.toString()).toBe('/abc/');
    });

    test('splits words', () => {
        expect(splitCamelCaseWord('hello')).toEqual(['hello']);
        expect(splitCamelCaseWord('helloThere')).toEqual(['hello', 'There']);
        expect(splitCamelCaseWord('HelloThere')).toEqual(['Hello', 'There']);
        expect(splitCamelCaseWord('BigÁpple')).toEqual(['Big', 'Ápple']);
        expect(splitCamelCaseWord('ASCIIToUTF16')).toEqual(['ASCII', 'To', 'UTF16']);
        expect(splitCamelCaseWord('URLsAndDBAs')).toEqual(['Urls', 'And', 'Dbas']);
        expect(splitCamelCaseWord('WALKingRUNning')).toEqual(['Walking', 'Running']);
    });

    test('extract word from text', () => {
        expect(Text.extractWordsFromText(`
            // could've, would've, couldn't've, wasn't, y'all, 'twas, shouldn’t
        `).toArray()).toEqual([
                { text: "could've", offset: 16 },
                { text: "would've", offset: 26 },
                { text: "couldn't've", offset: 36 },
                { text: "wasn't", offset: 49 },
                { text: "y'all", offset: 57 },
                { text: 'twas', offset: 65 },
                { text: 'shouldn’t', offset: 71 },
            ]);
    });

    test('extract words', () => {
        expect(Text.extractWordsFromText(`
            expect(splitCamelCaseWord('hello')).to.deep.equal(['hello']);
        `).toArray()).toEqual([
                { text: 'expect', offset: 13 },
                { text: 'splitCamelCaseWord', offset: 20 },
                { text: 'hello', offset: 40 },
                { text: 'to', offset: 49 },
                { text: 'deep', offset: 52 },
                { text: 'equal', offset: 57 },
                { text: 'hello', offset: 65 },
            ]);
        expect(Text.extractWordsFromText(`
            expect(splitCamelCaseWord('hello')).to.deep.equal(['hello']);
        `).toArray()).toEqual([
                { text: 'expect', offset: 13 },
                { text: 'splitCamelCaseWord', offset: 20 },
                { text: 'hello', offset: 40 },
                { text: 'to', offset: 49 },
                { text: 'deep', offset: 52 },
                { text: 'equal', offset: 57 },
                { text: 'hello', offset: 65 },
            ]);
        expect(Text.extractWordsFromText(`
            expect(splitCamelCaseWord('hello'));
        `).toArray()).toEqual([
                { text: 'expect', offset: 13 },
                { text: 'splitCamelCaseWord', offset: 20 },
                { text: 'hello', offset: 40 },
            ]);
    });

    test('extract words from code', () => {
        expect(Text.extractWordsFromCode(`
            expect(splitCamelCaseWord('hello')).to.deep.equal(['hello']);
        `).toArray()).toEqual([
                { text: 'expect', offset: 13 },
                { text: 'split', offset: 20 },
                { text: 'Camel', offset: 25 },
                { text: 'Case', offset: 30 },
                { text: 'Word', offset: 34 },
                { text: 'hello', offset: 40 },
                { text: 'to', offset: 49 },
                { text: 'deep', offset: 52 },
                { text: 'equal', offset: 57 },
                { text: 'hello', offset: 65 },
            ]);
        expect(Text.extractWordsFromCode(`
            expect(regExp.match(first_line));
        `).toArray()).toEqual([
                { text: 'expect', offset: 13 },
                { text: 'reg', offset: 20 },
                { text: 'Exp', offset: 23 },
                { text: 'match', offset: 27 },
                { text: 'first', offset: 33 },
                { text: 'line', offset: 39 },
            ]);
        expect(Text.extractWordsFromCode(`
            expect(aHELLO);
        `).toArray()).toEqual([
                { text: 'expect', offset: 13 },
                { text: 'a', offset: 20 },
                { text: 'HELLO', offset: 21 },
            ]);
        expect(Text.extractWordsFromCode(`
            const x = 'shouldn\'t';
        `).toArray()).toEqual([
                { text: 'const', offset: 13 },
                { text: 'x', offset: 19 },
                { text: 'shouldn\'t', offset: 24 },
            ]);
    });

    test('splits words like HTMLInput', () => {
        const words = Text.extractWordsFromCode('var value = HTMLInput.value;')
            .map(({text}) => text)
            .toArray();
        expect(words).toEqual(['var', 'value', 'HTML', 'Input', 'value']);
    });

    test('tests matchCase', () => {
        expect(Text.matchCase('Apple', 'orange')).toBe('Orange');
        expect(Text.matchCase('apple', 'ORANGE')).toBe('orange');
        expect(Text.matchCase('apple', 'orange')).toBe('orange');
        expect(Text.matchCase('APPLE', 'orange')).toBe('ORANGE');
        expect(Text.matchCase('ApPlE', 'OrangE')).toBe('OrangE');
        expect(Text.matchCase('apPlE', 'OrangE')).toBe('orangE');
    });

    test('tests skipping Chinese characters', () => {
        expect(Text.extractWordsFromCode(`
            <a href="http://www.ctrip.com" title="携程旅行网">携程旅行网</a>
        `).map(wo => wo.text).toArray()).toEqual(['a', 'href', 'http', 'www', 'ctrip', 'com', 'title', 'a']);
    });

    test('tests skipping Japanese characters', () => {
        expect(Text.extractWordsFromCode(`
            Example text: gitのpackageのみ際インストール
            gitのpackageのみ際インストール
            title="携程旅行网"
        `).map(wo => wo.text).toArray()).toEqual(['Example', 'text', 'git', 'package', 'git', 'package', 'title']);
    });

    test('tests Greek characters', () => {
        expect(Text.extractWordsFromCode(`
            Γ γ	gamma, γάμμα
        `).map(wo => wo.text).toArray()).toEqual(['Γ', 'γ', 'gamma', 'γάμμα']);
    });

    test('test case of Chinese characters', () => {
        expect(Text.isUpperCase('携程旅行网')).toBe(false);
        expect(Text.isLowerCase('携程旅行网')).toBe(false);
    });

    test('tests breaking up text into lines', () => {
        const parts = [
            '',
            '/*',
            ' * this is a comment.\r',
            ' */',
            '',
        ];
        const sampleText = parts.join('\n');
        const r = Text.extractLinesOfText(sampleText).map(a => a.text).toArray();
        expect(r.join('')).toBe(parts.join('\n'));
        const lines = [...Text.extractLinesOfText(sampleCode)].map(m => m.text);
        expect(lines.length).toBe(sampleCode.split('\n').length);
    });

    test('tests breaking up text into lines (single line)', () => {
        const parts = ['There is only one line.'];
        const sampleText = parts.join('\n');
        const r = Text.extractLinesOfText(sampleText).map(a => a.text).toArray();
        const rText = r.join('');
        expect(rText).toBe(parts.join('\n'));
        expect(rText).toBe(sampleText);
    });

    test('tests extractLinesOfText', () => {
        const linesA = [...Text.extractLinesOfText(sampleCode)].map(m => m.text);
        const linesB = Text.extractLinesOfText(sampleCode)
            .map(m => m.text)
            .toArray();
        expect(linesB).toEqual(linesA);
    });

    test('test extractText', () => {
        const line = Text.textOffset('This is a line of text to be processed.');
        const words = Text.extractWordsFromTextOffset(line);
        const results = words.map(wo => Text.extractText(line, wo.offset, wo.offset + wo.text.length)).toArray();
        const expected = words.map(wo => wo.text).toArray();
        expect(results).toEqual(expected);
    });
});

describe('Test the text matching functions', () => {
    test('isUpperCase', () => {
        expect(Text.isUpperCase('first')).toBe(false);
        expect(Text.isUpperCase('First')).toBe(false);
        expect(Text.isUpperCase('FIRST')).toBe(true);
    });
    test('isLowerCase', () => {
        expect(Text.isLowerCase('first')).toBe(true);
        expect(Text.isLowerCase('First')).toBe(false);
        expect(Text.isLowerCase('FIRST')).toBe(false);
    });
    test('isFirstCharacterUpper', () => {
        expect(Text.isFirstCharacterUpper('first')).toBe(false);
        expect(Text.isFirstCharacterUpper('First')).toBe(true);
        expect(Text.isFirstCharacterUpper('FIRST')).toBe(true);
    });
    test('isFirstCharacterLower', () => {
        expect(Text.isFirstCharacterLower('first')).toBe(true);
        expect(Text.isFirstCharacterLower('First')).toBe(false);
        expect(Text.isFirstCharacterLower('FIRST')).toBe(false);
    });
    // cSpell:ignore áello firstname
    test('ucFirst', () => {
        expect(Text.ucFirst('hello')).toBe('Hello');
        expect(Text.ucFirst('Hello')).toBe('Hello');
        expect(Text.ucFirst('áello')).toBe('Áello');
    });
    test('lcFirst', () => {
        expect(Text.lcFirst('hello')).toBe('hello');
        expect(Text.lcFirst('Hello')).toBe('hello');
        expect(Text.lcFirst('áello')).toBe('áello');
        expect(Text.lcFirst('Áello')).toBe('áello');
    });
    test('snakeToCamel', () => {
        expect(Text.snakeToCamel('first')).toBe('First');
        expect(Text.snakeToCamel('firstName')).toBe('FirstName');
        expect(Text.snakeToCamel('first_name')).toBe('FirstName');
        expect(Text.snakeToCamel('FIRST_NAME')).toBe('FIRSTNAME');
    });
    test('camelToSnake', () => {
        expect(Text.camelToSnake('first')).toBe('first');
        expect(Text.camelToSnake('firstName')).toBe('first_name');
        expect(Text.camelToSnake('first_name')).toBe('first_name');
        expect(Text.camelToSnake('FIRSTName')).toBe('first_name');
        expect(Text.camelToSnake('FIRSTNAME')).toBe('firstname');
    });
});


describe('Validates offset conversions', () => {
    function* getOffsets(haystack: string, needle: string) {
        let offset = -1;
        do {
            offset = haystack.indexOf(needle, offset + 1);
            if (offset > 0) {
                yield { offset, text: needle } as Text.TextOffset;
            } else {
                break;
            }
        } while (true);
    }

    test('calculateTextDocumentOffsets', () => {
        const offsets = [...getOffsets(sampleCode, 'const')];
        const results = Text.calculateTextDocumentOffsets('uri', sampleCode, offsets);
        expect(Object.keys(results)).not.toHaveLength(0);
        expect(results[0].row).toBe(6);
        expect(results[0].doc).toBe(sampleCode);
        expect(results[0].col).toBe(1);
    });
});

const sampleCode = `
/*
 * this is a comment.\r
 */

const text = 'some nice text goes here';
const url = 'https://www.google.com?q=typescript';
const url2 = 'http://www.weirddomain.com?key=jdhehdjsiijdkejshaijncjfhe';
const cssHexValue = '#cccd';
const cHexValue = 0x5612abcd;
const cHexValueBadCoffee = 0xbadc0ffee;

// spell-checker:disable
const unicodeHexValue = '\\uBABC';
const unicodeHexValue2 = '\\x\{abcd\}';

// spell-checker:enable

/* More code and comments */

// Make sure /* this works.

/* spell-checker:disable */

// nested disabled checker is not supported.

// spell-checker:disable

// nested spell-checker:enable <--> checking is now turned on.

// This will be checked

/*
 * spell-checker:enable  <-- this makes no difference because it was already turned back on.
 */

let text = '';
for (let i = 0; i < 99; ++i) {
    text += ' ' + i;
}

const string1 = 'This is a single quote string.  it\'s a lot of fun.'
const string2 = "How about a double quote string?";
const templateString = \`
can contain " and '

 \`;

$phpHereDocString = <<<SQL
    SELECT * FROM users WHERE id in :ids;
SQL;

$phpHereDocString = <<<"SQL"
    SELECT * FROM users WHERE id in :ids;
SQL;

$phpNowDocString = <<<'SQL'
    SELECT * FROM users WHERE id in :ids;
SQL;

// cSpell:disable

Not checked.

`;
