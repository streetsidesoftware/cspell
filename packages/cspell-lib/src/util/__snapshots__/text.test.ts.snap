// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validates offset conversions calculateTextDocumentOffsets for const 1`] = `
[
  {
    "col": 1,
    "line": {
      "offset": 32,
      "text": "const text = 'some nice text goes here';
",
    },
    "offset": 32,
    "row": 6,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 73,
      "text": "const url = 'https://www.google.com?q=typescript';
",
    },
    "offset": 73,
    "row": 7,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 124,
      "text": "const url2 = 'http://www.weirddomain.com?key=jdhehdjsiijdkejshaijncjfhe';
",
    },
    "offset": 124,
    "row": 8,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 198,
      "text": "const cssHexValue = '#cccd';
",
    },
    "offset": 198,
    "row": 9,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 227,
      "text": "const cHexValue = 0x5612abcd;
",
    },
    "offset": 227,
    "row": 10,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 257,
      "text": "const cHexValueBadCoffee = 0xbadc0ffee;
",
    },
    "offset": 257,
    "row": 11,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 323,
      "text": "const unicodeHexValue = '\\uBABC';
",
    },
    "offset": 323,
    "row": 14,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 357,
      "text": "const unicodeHexValue2 = '\\x{abcd}';
",
    },
    "offset": 357,
    "row": 15,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 840,
      "text": "const string1 = 'This is a single quote string.  it\\'s a lot of fun.'
",
    },
    "offset": 840,
    "row": 42,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 910,
      "text": "const string2 = "How about a double quote string?";
",
    },
    "offset": 910,
    "row": 43,
    "text": "const",
    "uri": "file:///uri",
  },
  {
    "col": 1,
    "line": {
      "offset": 962,
      "text": "const templateString = \`
",
    },
    "offset": 962,
    "row": 44,
    "text": "const",
    "uri": "file:///uri",
  },
]
`;

exports[`Validates offset conversions calculateTextDocumentOffsets for disable 1`] = `
[
  {
    "col": 18,
    "line": {
      "offset": 298,
      "text": "// spell-checker:disable
",
    },
    "offset": 315,
    "row": 13,
    "text": "disable",
    "uri": "file:///uri",
  },
  {
    "col": 18,
    "line": {
      "offset": 479,
      "text": "/* spell-checker:disable */
",
    },
    "offset": 496,
    "row": 23,
    "text": "disable",
    "uri": "file:///uri",
  },
  {
    "col": 11,
    "line": {
      "offset": 508,
      "text": "// nested disabled checker is not supported.
",
    },
    "offset": 518,
    "row": 25,
    "text": "disable",
    "uri": "file:///uri",
  },
  {
    "col": 18,
    "line": {
      "offset": 554,
      "text": "// spell-checker:disable
",
    },
    "offset": 571,
    "row": 27,
    "text": "disable",
    "uri": "file:///uri",
  },
  {
    "col": 11,
    "line": {
      "offset": 1241,
      "text": "// cSpell:disable
",
    },
    "offset": 1251,
    "row": 61,
    "text": "disable",
    "uri": "file:///uri",
  },
]
`;

exports[`Validates offset conversions calculateTextDocumentOffsets for text 1`] = `
[
  {
    "col": 7,
    "line": {
      "offset": 32,
      "text": "const text = 'some nice text goes here';
",
    },
    "offset": 38,
    "row": 6,
    "text": "text",
    "uri": "file:///uri",
  },
  {
    "col": 25,
    "line": {
      "offset": 32,
      "text": "const text = 'some nice text goes here';
",
    },
    "offset": 56,
    "row": 6,
    "text": "text",
    "uri": "file:///uri",
  },
  {
    "col": 5,
    "line": {
      "offset": 770,
      "text": "let text = '';
",
    },
    "offset": 774,
    "row": 37,
    "text": "text",
    "uri": "file:///uri",
  },
  {
    "col": 5,
    "line": {
      "offset": 816,
      "text": "    text += ' ' + i;
",
    },
    "offset": 820,
    "row": 39,
    "text": "text",
    "uri": "file:///uri",
  },
]
`;
